### PR TITLE
Added 'forward-key' option in outgoing kafka connector to forward rec…

### LINKED ISSUE #1301: Allow the Kafka outgoing connector to forward the incoming key
This feature would allow forwarding the key of incoming records into outgoing records automatically. 

When a message is going to be written to Kafka, and the key has not been explicitly set (neither in the record, in the outgoing metadata, nor in the connector configuration) and if there is an incoming Kafka metadata, this option would use the key from that incoming Kafka metadata.

### PR DESCRIPTION
**Pull request by ozangunalp**
…ord key from incoming message to the outgoing message

Fixes #1301

### CODE REVIEW

File: smallrye-reactive-messaging-kafka/src/test/java/io/smallrye/reactive/messaging/kafka/MetadataPropagationTest.java [hunk 1]
```diff
@@ -120,6 +140,27 @@ private KafkaMapBasedConfig getKafkaSinkConfigForMyAppProcessingData(String topi
         return builder.build();
     }
 
+    private KafkaMapBasedConfig getKafkaSinkConfigForAppProcessingDataForwardKey(String topicOut, String topicIn) {
+        KafkaMapBasedConfig.Builder builder = KafkaMapBasedConfig.builder();
+        builder.put("mp.messaging.outgoing.kafka.value.serializer", StringSerializer.class.getName());
```
Comment (Ladicek) [inline smallrye-reactive-messaging-kafka/src/test/java/io/smallrye/reactive/messaging/kafka/MetadataPropagationTest.java:145]: `IntegerSerializer`?

**Comment by Ladicek** [comment]
Looks OK other than that.

File: smallrye-reactive-messaging-kafka/src/test/java/io/smallrye/reactive/messaging/kafka/MetadataPropagationTest.java (hunk 1 above)
Comment (ozangunalp) [reply smallrye-reactive-messaging-kafka/src/test/java/io/smallrye/reactive/messaging/kafka/MetadataPropagationTest.java:145]: Well it doesn't make any difference for the test :)

File: smallrye-reactive-messaging-kafka/src/main/java/io/smallrye/reactive/messaging/kafka/KafkaConnector.java [hunk 2]
```diff
@@ -106,6 +106,7 @@
 @ConnectorAttribute(name = "cloud-events-mode", type = "string", direction = Direction.OUTGOING, description = "The Cloud Event mode (`structured` or `binary` (default)). Indicates how are written the cloud events in the outgoing record", defaultValue = "binary")
 @ConnectorAttribute(name = "close-timeout", type = "int", direction = Direction.OUTGOING, description = "The amount of milliseconds waiting for a graceful shutdown of the Kafka producer", defaultValue = "10000")
 @ConnectorAttribute(name = "merge", direction = OUTGOING, description = "Whether the connector should allow multiple upstreams", type = "boolean", defaultValue = "false")
+@ConnectorAttribute(name = "forward-key", direction = OUTGOING, description = "Forward incoming message key to the outgoing message key", type = "boolean", defaultValue = "false")
```
Comment (cescoffier) [inline smallrye-reactive-messaging-kafka/src/main/java/io/smallrye/reactive/messaging/kafka/KafkaConnector.java:109]: I'm not sure about the wording. Maybe something like: propagate-record-key.

File: smallrye-reactive-messaging-kafka/src/main/java/io/smallrye/reactive/messaging/kafka/impl/KafkaSink.java [hunk 3]
```diff
@@ -266,6 +267,11 @@ private Object getKey(Message<?> message,
             return ((Record) message.getPayload()).key();
         }
 
+        // Then, check if the message contains incoming metadata from which we can forward the key
+        if (configuration.getForwardKey() && message.getMetadata(IncomingKafkaRecordMetadata.class).isPresent()) {
+            return message.getMetadata(IncomingKafkaRecordMetadata.class).get().getKey();
```
Comment (cescoffier) [inline smallrye-reactive-messaging-kafka/src/main/java/io/smallrye/reactive/messaging/kafka/impl/KafkaSink.java:272]: What do we need to do if it's null. Should be used the configured one or null?

**Review by cescoffier — CHANGES_REQUESTED** [review]

**Comment by cescoffier** [comment]
This feature should be documented explicitly

File: smallrye-reactive-messaging-kafka/src/main/java/io/smallrye/reactive/messaging/kafka/impl/KafkaSink.java (hunk 3 above)
Comment (ozangunalp) [reply smallrye-reactive-messaging-kafka/src/main/java/io/smallrye/reactive/messaging/kafka/impl/KafkaSink.java:272]: My initial implementation did a fallback on configured `key` if both `forward-key` and `key` were configured and incoming record key was null (or `IncomingKafkaRecordMetadata` wasn't there).
I now think that it makes more sense.

File: smallrye-reactive-messaging-kafka/src/main/java/io/smallrye/reactive/messaging/kafka/KafkaConnector.java (hunk 2 above)
Comment (ozangunalp) [reply smallrye-reactive-messaging-kafka/src/main/java/io/smallrye/reactive/messaging/kafka/KafkaConnector.java]: I changed the config name to `propagate-record-key`.

**Comment by ozangunalp** [comment]
@cescoffier I added a short section inside kafka outbound documentation.

File: documentation/src/main/doc/modules/kafka/pages/outbound.adoc [hunk 4]
```diff
@@ -79,6 +79,24 @@ For example, you can add Kafka headers, configure the record key...
...
+
+[source, java]
+----
+include::example$outbound/KafkaPriceProcessor.java[tag=code]
+----
+
+Enabled with `mp.messaging.outgoing.[channel-name].propagate-record-key=true` configuration,
+record key propagation produces the outgoing record with the same _key_ as the incoming record.
+
+If the outgoing record already contains a _key_, it *won't be overridden* by the incoming record key.
```
Comment (cescoffier) [inline documentation/src/main/doc/modules/kafka/pages/outbound.adoc:97]: ** no ? * is just italic.

File: smallrye-reactive-messaging-kafka/src/test/java/io/smallrye/reactive/messaging/kafka/MetadataPropagationTest.java [hunk 5]
```diff
@@ -66,6 +68,48 @@ public void testFromKafkaToAppToKafka() {
...
+        });
+
+        await().until(() -> messages.size() >= 10);
+
+        assertThat(messages).containsKeys("0", "1", "2", "3", "4", "5", "6", "7", "8", "9");
+        assertThat(messages).containsValues("1", "2", "3", "4", "5", "6", "7", "8", "9", "10");
+    }
+
+    @Test
+    public void testFromKafkaToAppToKafkaForwardKeyReturningMessage() {
```
Comment (cescoffier) [inline smallrye-reactive-messaging-kafka/src/test/java/io/smallrye/reactive/messaging/kafka/MetadataPropagationTest.java:91]: Can you add a test verifying when the incoming key is `null`?

**Review by cescoffier — CHANGES_REQUESTED** [review]

File: documentation/src/main/doc/modules/kafka/pages/outbound.adoc (hunk 4 above)
Comment (ozangunalp) [reply documentation/src/main/doc/modules/kafka/pages/outbound.adoc:97]: Wait you got me thinking :)

In asciidoc
```asciidoc
*bold* or **bold**text
_italic_ or __italic__text
```
In markdown
```markdown
**bold** or __bold__
*italic* or _italic_
```

Comment (cescoffier) [reply documentation/src/main/doc/modules/kafka/pages/outbound.adoc:97]: oh... that's not confusing at all :-D

File: smallrye-reactive-messaging-kafka/src/test/java/io/smallrye/reactive/messaging/kafka/MetadataPropagationTest.java (hunk 5 above)
Comment (ozangunalp) [reply smallrye-reactive-messaging-kafka/src/test/java/io/smallrye/reactive/messaging/kafka/MetadataPropagationTest.java:90]: Pushed the test

**Review by cescoffier — APPROVED** [review]
